[NC-661] add column chart native widget

Ahmad-Elsayed opened the pull request:
## Checklist
- Contains unit tests ✅ 
- Contains breaking changes ❌
- Contains Atlas changes ✅ 
- Compatible with: MX 9️⃣

#### Native specific
- Works in Android ✅ 
- Works in iOS ✅ 
- Works in Tablet ✅ 

#### Feature specific
- Comply with designs ✅ 
- Comply with PM's requirements ✅ 

## This PR contains
- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
_Add column chart native_

## What should be covered while testing?
_It should have the same functionality as Bar chart widget but with a vertical orientation_

Ahmad-Elsayed committed:
fix(column-chart-native): update icons

jordanch commented on packages/pluggableWidgets/column-chart-native/README.md:
```diff
@@ -0,0 +1,40 @@
+### Custom styling
+
+The look and feel of the chart is fully customizable. Check out the
+[Native Mobile Styling](TODO) documentation for a full overview of
```
heres a todo

jordanch commented on packages/pluggableWidgets/column-chart-native/package.json:
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "column-chart-native",
+  "widgetName": "ColumnChart",
+  "version": "2.1.0",
```
1.0.0?

jordanch commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,303 @@
+            return null;
+        }
+
+        // datum.y is actually the X axis data points due to the way Victory handles horizontal charts
```
Is this still relevant, as the chart is no longer horizontal?

jordanch commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,303 @@
+                                : null}
+                            <View onLayout={updateChartDimensions} style={{ flex: 1 }}>
+                                {chartDimensions ? (
+                                    // flip the domain padding x and y axis values due to how victory library handles
```
Is this still relevant, as the chart is no longer horizontal?

jordanch commented on packages/pluggableWidgets/column-chart-native/src/utils/__tests__/StyleUtils.spec.ts:
```diff
@@ -0,0 +1,212 @@
+        });
+    });
+
+    describe("mapToBarStyles", () => {
```
This is named incorrect although its not an issue as this code will be refactored.

jordanch reviewed (changes requested):
nice one, minor changes

Ahmad-Elsayed committed:
fix(column-chart-native): axis orientation

Ahmad-Elsayed committed:
fix(column-chart-native): domain padding parameters

Ahmad-Elsayed committed:
fix(column-chart-native): add structure preview

Ahmad-Elsayed committed:
fix(column-chart-native): add victory-bar dep

Ahmad-Elsayed committed:
chore: retrigger actions

Ahmad-Elsayed committed:
fix(column-chart-native): series loader issue

Ahmad-Elsayed committed:
fix(column-chart-native): update detox dependency

UrazAkgultan reviewed (dismissed):
LGTM 👍

pratikgoyal9 reviewed (changes requested):
Need to resolve the issues reported in the test report.

Ahmad-Elsayed committed:
feat(column-chart-native): add widget

Ahmad-Elsayed committed:
fix(column-chart-native): fix issues

UrazAkgultan commented on packages/theming/atlas/src/theme/native/custom-variables.ts:
```diff
@@ -1,3 +1,4 @@
+import { VariablesColumnChart } from "./../../themesource/atlas_core/native/types/variables.d";
```
This import is defined twice. You can move the `VariablesColumnChart` to the below import.

Ahmad-Elsayed replied:
done

UrazAkgultan commented on packages/pluggableWidgets/column-chart-native/package.json:
```diff
@@ -0,0 +1,36 @@
+    "@mendix/pluggable-widgets-tools": ">=8.9.2",
+    "detox": "^19.5.7",
+    "eslint": "^7.20.0",
+    "victory-bar": "35.11.4"
```
You don't need to add `victory-bar` here.

Ahmad-Elsayed replied:
done

UrazAkgultan reviewed (commented):
I left some comments 👍

Ahmad-Elsayed committed:
fix(column-chart-native): remove unused dependency

UrazAkgultan reviewed (dismissed):
Good effort, thanks 👍

Ahmad-Elsayed committed:
fix(column-chart-native): add todo comments

Ahmad-Elsayed committed:
fix(column-chart-native): handle series sorting

fix(column-chart-native): handle sorting

fix(column-chart-native): add defaults for missing keys before sorting

fix(column-chart-native): upadate column series description

tamelbrun commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+
+    const warningMessagePrefix = useMemo(() => (warningPrefix ? warningPrefix + "i" : "I"), [warningPrefix]);
+
+    const dataTypesResult = useMemo(() => getDataTypes(sortedSeries), [sortedSeries]);
```
It does not matter whether you do this on the sorted series or the original one, right? In that case, if you move this to the beginning, you already know the data types and don't have to iterate over the data points again to check if they are sortable.

Ahmad-Elsayed replied:
done

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+        const isNotSortable = series.some(({ dataPoints }) =>
+            dataPoints.some(({ x, y }) => typeof x !== "string" || typeof y !== "number")
+        );
+        return isNotSortable ? series : sortSeriesDataPoints(series, sortOrder);
```
done

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+}: ColumnChartProps): ReactElement | null {
+    const sortedSeries = useMemo(() => {
+        const isNotSortable = series.some(({ dataPoints }) =>
+            dataPoints.some(({ x, y }) => typeof x !== "string" || typeof y !== "number")
```
the library itself doesn't allow sorting numbers and dates keys

Ahmad-Elsayed replied:
https://formidable.com/open-source/victory/docs/victory-bar/

tamelbrun replied:
But you're now pre-sorting the data here instead of letting the library handle the sorting, right? If so, does it still matter what the library allows?

"We don't need this feature for numbers and dates" is also a valid answer, btw :P.

tamelbrun replied:
But you're now pre-sorting the data here instead of letting the library handle the sorting, right? If so, does it still matter what the library allows?

"We don't need this feature for numbers and dates" is also a valid answer, btw :P.

Ahmad-Elsayed replied:
if I sorted the keys in respect to their values the result will be something like this [{x: 10, y:1}, {x: 2, y: 3}, {x:4, y: 5}]
if the data is in a stack, the resulted chart will be missing a lot of data. this is a library issue

Ahmad-Elsayed replied:
if the numeric or date keys are in correct order, it is works fine.

tamelbrun replied:
(It actually looks like the current implementation handles this correctly, though? :))

Ahmad-Elsayed replied:
yes, I have added some code to sort the keys in an ascending order in case the key are numbers or dates to prevent this issue

tamelbrun replied:
Then I will gladly resolve this one! 😄

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+            })
+            .map(key => ({
+                x: key,
+                y: dataPoints.find((point: ColumnDataPoint<string, number>) => point.x === key)?.y || 0
```
I did that to make all the series have the same keys, because of a library issue.

tamelbrun replied:
Ah, that's unfortunate. But the `?` also suddenly makes sense to me now 😅.

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+
+    return series.map(seriesItem => {
+        const dataPoints = seriesItem.dataPoints as Array<ColumnDataPoint<string, number>>;
+        seriesItem.dataPoints = Object.keys(keysSum)
```
done

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+                const ColumnColor = ColumnColorPalette?.[index] || "black";
+
+                if (ColumnColorPalette) {
+                    index = index + 1 === ColumnColorPalette.length ? 0 : index + 1;
```
done

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+        ) {
+            if (extractedXAxisLabelStyle.relativePositionGrid !== undefined) {
+                console.warn(
+                    `${warningMessagePrefix}nvalid value for X axis label style property, relativePositionGrid, valid values are "bottom" and "right".`
```
done

tamelbrun replied:
It looks like there will be two spaces now, one from the warning prefix and one here.

Ahmad-Elsayed replied:
done

Ahmad-Elsayed committed:
fix(column-chart-native): handle sorting date and number datatypes and code refactor

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+    const keysSum: { [key: string]: number } = {};
+    series.forEach(({ dataPoints }) => {
+        dataPoints.forEach(({ x, y }) => {
+            const key = x.toString();
```
done

tamelbrun commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,367 @@
+function getDataTypes(series: ColumnChartSeries[]): DataTypeResult | Error | undefined {
+    let dataTypes: DataTypeResult | undefined;
+
+    for (const _series of series) {
+        const { dataPoints } = _series;
```
Let's avoid the `_` by calling this variable `seriesItem` like everywhere else? :)

Ahmad-Elsayed replied:
done

tamelbrun commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,367 @@
+}: ColumnChartProps): ReactElement | null {
+    const dataTypesResult = useMemo(() => getDataTypes(series), [series]);
+    const sortedSeries = useMemo(() => {
+        const seriesDataType = dataTypesResult as DataTypeResult | undefined;
```
This cast is unsafe. If the `sortSeriesDataPoints` function does not crash when the result is actually an `Error`, that is by sheer luck. Instead, I would immediately return the unsorted `series` if `dataTypesResult` is either an `Error` or `undefined`, as you cannot sort in those cases. TypeScript will automatically narrow the variable to `DataTypeResult` afterwards, so `sortSeriesDataPoints` would not need to deal with `undefined` anymore either.

Ahmad-Elsayed replied:
done

tamelbrun commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+    // Column Chart user-styling may be missing for certain series. A palette is passed, any missing colours
+    // fallback to a colour from the palette.
+    const normalizedColumnColors: string[] = useMemo(() => {
+        const ColumnColorPalette = style.columns?.columnColorPalette?.split(";");
```
This one is fixed now, but the other ones are not ;).

Ahmad-Elsayed replied:
which ones do you mean?

tamelbrun replied:
`Columns` and `ColumnStyles` still start with uppercase.

Ahmad-Elsayed replied:
done

tamelbrun commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,367 @@
+}
+
+const getScale = (dataTypesResult: string): "linear" | "time" | undefined => {
+    if (dataTypesResult === "number" || dataTypesResult === "string") {
+        return "linear";
+    } else if (dataTypesResult === "object") {
+        return "time";
+    }
+
+    return undefined;
```
I would personally go for a `switch` here, but I'll leave that up to you :).

Ahmad-Elsayed replied:
done

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,367 @@
+        dataPoints.forEach(({ x, y }) => {
+            const key = x as string;
+            if (key in keysSum) {
+                keysSum[key] += Number(y);
+            } else {
+                keysSum[key] = Number(y);
```
done

Ahmad-Elsayed committed:
fix(column-chart-native): code refactor and change wrong variable names

Ahmad-Elsayed commented on packages/pluggableWidgets/column-chart-native/src/components/ColumnChart.tsx:
```diff
@@ -0,0 +1,328 @@
+            }
+
+            extractedYAxisLabelStyle.relativePositionGrid = "top";
+        }
```
it is checking for X label and Y label. I updated it to be more readable.

tamelbrun replied:
Ah, I see. If you add a newline before the first `if`, I can live with it :). But my suggestion was something like this (written from the top of my head 🤞🏻):

```
function validateRelativePositionGrid(style: { relativePositionGrid: string }, axis: "X" | "Y", ...validValues: string[]) {
	const position = style.relativePositionGrid;
	
	if (!validValues.includes(position)) {
		if (position !== undefined) {
			console.warn(
			   `${warningPrefix}Invalid value for ${axis} axis label style property relativePositionGrid, valid values are: ${validValues.join(", ")}`
			);
		}
		style.relativePositionGrid = validValues[0];
	}
}
```

and then

```
checkRelativePositionGrid(extractedXAxisLabelStyle, "X", "bottom", "right");
checkRelativePositionGrid(extractedYAxisLabelStyle, "Y", "top", "left");
```

Ahmad-Elsayed replied:
done

tamelbrun replied:
✨

Ahmad-Elsayed committed:
fix(column-chart-native): update legend assets, refactor axis label styles

Ahmad-Elsayed committed:
fix(column-chart-native): grid styles defaults

Ahmad-Elsayed committed:
fix(column-chart-native): handle 0 label column
